README.rst
MapServer
=========

| |Build Status| |Appveyor Build Status| |Coveralls Status|

-------
Summary
-------

Test
   
MapServer is a system for developing web-based GIS applications. 
The basic system consists of a CGI program that can be configured to 
respond to a variety of spatial requests like making maps, scalebars, 
and point, area and feature queries. Virtually all aspects of an 
application, from web interface to map appearance can be developed 
without any programming. For the more ambitious user, MapServer 
applications can be enhanced using Python, PHP, Java, JavaScript or 
many other web technologies. For more  information and complete 
documentation please visit:

  https://mapserver.org/

Bug reports and enhancement submissions can be reported in the MapServer 
issue tracker at the following url.   If you do make changes and/or enhancements, 
please let us know so that they might be incorporated into future releases.

  https://github.com/MapServer/MapServer/issues


Join the MapServer user mailing list online at:

  https://mapserver.org/community/lists.html

 

Credits
-------

MapServer was originally written by Stephen Lime. Major funding for development of 
MapServer has been provided by NASA through cooperative argreements with 
the University of Minnesota, Department of Forest Resources.

PHP/MapScript developed by DM Solutions Group.

GDAL/OGR support and significant WMS support provided by DM Solutions Group 
which received funding support from Canadian Government's GeoConnections 
Program and the Canadian Forest Service.

Raster support developed by Pete Olson of the State of Minnesota, Land 
Management Information Center, and maintained by Frank Warmerdam (DM 
Solutions).

PostGIS spatial database support provided by Dave Blasby of Refractions 
Research.

PDF support developed by Jeff Spielberg and Jamie Wall of Market Insite Group, 
Inc.

OracleSpatial support developed by Rodrigo Cabral of CTTMAR/UNIVALI, Brazil.

Portions Copyright (c) 1998 State of Minnesota, Land Management Information 
Center.

Portions derived from Shapelib, Copyright 1995-1999 Frank Warmerdam.

Supporting packages are covered by their own copyrights.

License
-------

::

  Copyright (c) 2008-2020 Open Source Geospatial Foundation.
  Copyright (c) 1996-2008 Regents of the University of Minnesota.

  Permission is hereby granted, free of charge, to any person obtaining a copy 
  of this software and associated documentation files (the "Software"), to deal 
  in the Software without restriction, including without limitation the rights 
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell 
  copies of the Software, and to permit persons to whom the Software is furnished
  to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in all 
  copies of this Software or works derived from this Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE 
  SOFTWARE.


.. |Build Status| image:: https://travis-ci.com/MapServer/MapServer.svg?branch=master
   :target: https://travis-ci.com/MapServer/MapServer

.. |Appveyor Build Status| image:: https://ci.appveyor.com/api/projects/status/vw1n07095a8bg23u?svg=true
   :target: https://ci.appveyor.com/project/mapserver/mapserver

.. |Coveralls Status| image:: https://coveralls.io/repos/github/mapserver/mapserver/badge.svg?branch=master
   :target: https://coveralls.io/github/mapserver/mapserver?branch=master


debian/changelog
mapserver (7.6.2-inspire-1~bpo10+1) UNRELEASED; urgency=medium

  * New release with INSPIRE metadata extensions

 -- Florian Esser <fe@wetransform.to>  Tue, 05 Jan 2021 21:19:48 +0100

mapserver (7.6.2-1~bpo10+1) buster-backports; urgency=medium

  * Rebuild for buster-backports.

 -- Bas Couwenberg <sebastic@debian.org>  Mon, 14 Dec 2020 06:02:08 +0100

mapserver (7.6.2-1) unstable; urgency=medium

  * Update symbols for other architectures.
  * Move from experimental to unstable.

 -- Bas Couwenberg <sebastic@debian.org>  Wed, 09 Dec 2020 06:01:02 +0100

mapserver (7.6.2-1~exp1) experimental; urgency=medium

  * New upstream release.
  * Update symbols for other architectures.
  * Bump watch file version to 4.
  * Update lintian overrides.
  * Bump Standards-Version to 4.5.1, no changes.
  * Require at least cmake 3.0.
  * Update copyright file.
  * Update upstream metadata.
  * Update Homepage URL to use HTTPS.
  * Update symbols for amd64.

 -- Bas Couwenberg <sebastic@debian.org>  Tue, 08 Dec 2020 05:29:11 +0100

mapserver (7.6.1-1~bpo10+1) buster-backports; urgency=medium

  * Rebuild for buster-backports.

 -- Bas Couwenberg <sebastic@debian.org>  Thu, 06 Aug 2020 05:35:52 +0200

mapserver (7.6.1-1) unstable; urgency=medium

  * Update symbols for other architectures.
  * Move from experimental to unstable.

 -- Bas Couwenberg <sebastic@debian.org>  Sat, 01 Aug 2020 08:05:35 +0200

mapserver (7.6.1-1~exp1) experimental; urgency=medium

  * New upstream release.
  * Drop phpng-mapscript.patch, included upstream.
  * Update symbols for amd64.
  * Add lintian overrides for file-references-package-build-path.

 -- Bas Couwenberg <sebastic@debian.org>  Fri, 31 Jul 2020 19:35:49 +0200

mapserver (7.6.0-2~bpo10+1) buster-backports; urgency=medium

  * Rebuild for buster-backports.

 -- Bas Couwenberg <sebastic@debian.org>  Fri, 03 Jul 2020 07:20:13 +0200

mapserver (7.6.0-2) unstable; urgency=medium

  * Add upstream patch to install mapscript.php.
    (LP: #1884854)
  * Include mapscript.php in php-mapscript-ng.
  * Add lintian overrides for breakout-link.

 -- Bas Couwenberg <sebastic@debian.org>  Sat, 27 Jun 2020 20:32:54 +0200

mapserver (7.6.0-1~bpo10+1) buster-backports; urgency=medium

  * Rebuild for buster-backports.

 -- Bas Couwenberg <sebastic@debian.org>  Thu, 14 May 2020 06:25:59 +0200

mapserver (7.6.0-1) unstable; urgency=medium

  * New upstream release.
  * Update 7.6.0~rc4 symbols for other architectures.
  * Strip pre-releases from symbols version.
  * Move from experimental to unstable.

 -- Bas Couwenberg <sebastic@debian.org>  Fri, 08 May 2020 17:38:17 +0200

mapserver (7.6.0~rc4-1~exp1) experimental; urgency=medium

  * New upstream release candidate.
  * Drop override for mailing-list-obsolete-in-debian-infrastructure.
  * Update symbols for other architectures.

 -- Bas Couwenberg <sebastic@debian.org>  Tue, 05 May 2020 05:46:51 +0200

mapserver (7.6.0~rc3-1~exp1) experimental; urgency=medium

  * New upstream release candidate.
  * Update symbols for amd64.
  * Add override for mailing-list-obsolete-in-debian-infrastructure.

 -- Bas Couwenberg <sebastic@debian.org>  Sat, 25 Apr 2020 06:45:11 +0200

mapserver (7.6.0~rc2-1~exp1) experimental; urgency=medium

  * New upstream release candidate.

 -- Bas Couwenberg <sebastic@debian.org>  Sat, 11 Apr 2020 06:47:59 +0200

mapserver (7.6.0~rc1-1~exp1) experimental; urgency=medium

  * New upstream release candidate.

 -- Bas Couwenberg <sebastic@debian.org>  Mon, 06 Apr 2020 05:56:37 +0200

mapserver (7.6.0~beta2-1~exp1) experimental; urgency=medium

  * New upstream beta release.
  * Drop -D_GNU_SOURCE, fixed upstream.

 -- Bas Couwenberg <sebastic@debian.org>  Sun, 29 Mar 2020 08:46:12 +0200

mapserver (7.6.0~beta1-1~exp1) experimental; urgency=medium

  * New upstream beta release.
  * Drop obsolete CMake options.
  * Update copyright years for copyright holders.
  * Drop swig-4.0.1.patch, included upstream.
  * Add -D_GNU_SOURCE to CPPFLAGS.
  * Update symbols for amd64.

 -- Bas Couwenberg <sebastic@debian.org>  Mon, 23 Mar 2020 06:51:15 +0100

mapserver (7.4.4-1~bpo10+1) buster-backports; urgency=medium

  * Rebuild for buster-backports.

 -- Bas Couwenberg <sebastic@debian.org>  Fri, 27 Mar 2020 18:00:55 +0100

mapserver (7.4.4-1) unstable; urgency=high

  * New upstream release.
  * Drop mapscript-buffer-overflow.patch, included upstream.

 -- Bas Couwenberg <sebastic@debian.org>  Fri, 20 Mar 2020 21:15:45 +0100

mapserver (7.4.3-3) unstable; urgency=high

  * Bump debhelper compat to 10, changes:
    - Drop --parallel option, enabled by default
  * Add upstream patch to fix PHPMapScript error handling vulnerabilities.

 -- Bas Couwenberg <sebastic@debian.org>  Fri, 20 Mar 2020 20:40:37 +0100

mapserver (7.4.3-2) unstable; urgency=medium

  * Bump Standards-Version to 4.5.0, no changes.
  * Add upstream patch to fix FTBFS with SWIG 4.0.1.
    (closes: #951886)

 -- Bas Couwenberg <sebastic@debian.org>  Sat, 22 Feb 2020 19:25:03 +0100

mapserver (7.4.3-1~bpo10+2) buster-backports; urgency=high

  * Add upstream patch to fix PHPMapScript error handling vulnerabilities.

 -- Bas Couwenberg <sebastic@debian.org>  Fri, 20 Mar 2020 20:53:42 +0100

mapserver (7.4.3-1~bpo10+1) buster-backports; urgency=medium

  * Rebuild for buster-backports.

 -- Bas Couwenberg <sebastic@debian.org>  Sun, 22 Dec 2019 06:37:42 +0100

mapserver (7.4.3-1) unstable; urgency=medium

  * New upstream release.
  * Bump Standards-Version to 4.4.1, no changes.
  * Drop Provides: ${python3:Provides}.
  * Change python3-all{,-dev} build dependency to python3{,-dev},
    only the default interpreter is used.
  * Drop Name field from upstream metadata.
  * Drop unused overrides for file-references-package-build-path.
  * Update symbols for amd64.
  * Rename TODO.Debian to TODO.

 -- Bas Couwenberg <sebastic@debian.org>  Tue, 17 Dec 2019 05:34:32 +0100

mapserver (7.4.2-1~bpo10+1) buster-backports; urgency=medium


epplib.c
/******************************************************************************
 *
 * Project:  MapServer
 * Purpose:  Low level EPP file access, supports drawEPP() in mapraster.c.
 * Author:   Steve Lime and the MapServer team.
 *
 ******************************************************************************
 * Copyright (c) 1996-2005 Regents of the University of Minnesota.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in 
 * all copies of this Software or works derived from this Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 ****************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "epplib.h"
#include "mapserver.h"

MS_CVSID("$Id: epplib.c 6428 2007-07-31 19:58:45Z dmorissette $")

static int REVERSE; /* set to 1 on bigendian machines */

void swap2(short *x, int count)
{
  int n;
  char tmp;
  for (n=0; n<count; n++) {
    tmp=*(char *)x;
    *(char *)x=*(((char *)x)+1);
    *(((char *)x)+1)=tmp;
    x++;
  }
}

void swap4(long *x, int count)
{
  int n;
  char tmp;
  for (n=0; n<count; n++) {
    tmp=*(char *)x;
    *(char *)x=*(((char *)x)+3);
    *(((char *)x)+3)=tmp;
    tmp=*(((char *)x)+1);
    *(((char *)x)+1)=*(((char *)x)+2);
    *(((char *)x)+2)=tmp;
    x++;
  }
}

void swap8(double *x, int count)
{
  int n;
  char tmp;
  for (n=0; n<count; n++) {
    tmp=*(char *)x;
    *(char *)x=*(((char *)x)+7);
    *(((char *)x)+7)=tmp;
    tmp=*(((char *)x)+1);
    *(((char *)x)+1)=*(((char *)x)+6);
    *(((char *)x)+6)=tmp;
    tmp=*(((char *)x)+2);
    *(((char *)x)+2)=*(((char *)x)+5);
    *(((char *)x)+5)=tmp;
    tmp=*(((char *)x)+3);
    *(((char *)x)+3)=*(((char *)x)+4);
    *(((char *)x)+4)=tmp;
    x++;
  }
}

static char getrow(eppfile *epp,rowptr buff)
{
  rowptr rp;
  char *cp;
  int col,ncol,nc,n,v;
  ncol=epp->lc-epp->fc+1;
  rp=buff+1;
  for (col=0; col<ncol; col+=nc) {
    if (epp->fp-epp->fptr>4095) {
      memcpy(epp->fptr,epp->fptr+4096,256);
      epp->fp-=4096;
      nc=fread(epp->fptr+256,1,4096,epp->fil);
      /* eof check ?? */
      }
    nc=*epp->fp++;
    if (nc!=0) {
      v=*epp->fp++;
      for (n=0; n<nc; n++) *rp++=v;
    } else {
      nc=*epp->fp++;
      for (n=0; n<nc; n++) *rp++=*epp->fp++;
    }
  }
  if (epp->kind==16 && col==ncol) {
    if (REVERSE) cp=((char *)(epp->rptr+1))+2;  /* pick up high half */
    else cp=((char *)(epp->rptr+1))+3;
    for (col=0; col<ncol; col+=nc) {
      if (epp->fp-epp->fptr>4095) {
	memcpy(epp->fptr,epp->fptr+4096,256);
	epp->fp-=4096;
	nc=fread(epp->fptr+256,1,4096,epp->fil);
	/* eof check ?? */
	}
      nc=*epp->fp++;
      if (nc!=0) {
	v=*epp->fp++;
	for (n=0; n<nc; n++) {
	  *cp=v;
	  cp+=4;
	}
      } else {
	nc=*epp->fp++;
	for (n=0; n<nc; n++) {
	  *cp=*epp->fp++;
	  cp+=4;
	}
      }
    }
  }
  return (ncol==col);
}

char get_row(eppfile *EPP)
{
char STR1[80];
  if (!getrow(EPP,EPP->rptr)) {
    sprintf(STR1,"Error in reading file %s",EPP->filname);
    msSetError(MS_IMGERR,STR1,"drawEPP()");
    eppclose(EPP);
    return 0;
  }
  EPP->row_cnt++;
  return 1;
}

char eppreset(eppfile *EPP)
{  /*Open_file*/
  int nrd;
  char STR1[102];

  { union { long i; char c[4]; } u; u.i=1; REVERSE=(u.c[0]==0); }
  EPP->fil=fopen(EPP->filname,"rb");
  if (EPP->fil==NULL) {
    sprintf(STR1,"Can't open %s",EPP->filname);
    msSetError(MS_IMGERR,STR1,"drawEPP()");
    return 0;
  }
  nrd=fread(&EPP->fr,1,128,EPP->fil);
      /*fr is start of data*/
  if (REVERSE) {        /* fix up header on big-endian machines */
    swap2(&EPP->fr,4);
    swap8(&EPP->fry,4);
    swap2((short *)&EPP->kind,4);
    swap8(&EPP->sfactor,1);
    swap4(&EPP->access_ptr,1);
    swap2((short *)&EPP->minval,2);
  }
  if (EPP->kind!=8 && (EPP->kind!=16 || nrd!=128)) {
    sprintf(STR1,"%s is not an EPPL file.",EPP->filname);
    msSetError(MS_IMGERR,STR1,"drawEPP()");
    fclose(EPP->fil);
    return 0;
  }
  EPP->row_cnt=0;
  if (EPP->kind==8 && (EPP->minval > 256 || EPP->maxval > 256)) {
    /*old EPPL file; try to fix up*/
    EPP->minval=0;
    EPP->maxval=255;
  }
  EPP->fptr =(unsigned char *)malloc(sizeof(file_buffer));
  nrd=fread(&EPP->fptr[384],1,3968,EPP->fil);
      /*so we read even 4096 byte chunks*/
  EPP->fp=EPP->fptr+384;
  if (nrd==3968)
    EPP->lastbyte=4096;
  else
    EPP->lastbyte=nrd + 384;
  EPP->rptr =(rowptr)malloc((EPP->lc-EPP->fc+3) * sizeof(*EPP->rptr));
  EPP->access_table=NULL;
  return 1;
}

static char epprewind(eppfile *EPP)
{
  if (!eppclose(EPP))
    return 0;
  return eppreset(EPP);
}

char position(eppfile *EPP,int row)
{
  int r,nread;
  long z,now;

  if (EPP->access_table==NULL) {  /*first time through*/
    EPP->access_table=(unsigned short *)malloc((EPP->lr-EPP->fr+2)*2+2);
	/*2 bytes of extra space*/
    EPP->saved=EPP->access_table;
    now=ftell(EPP->fil);
    fseek(EPP->fil,EPP->access_ptr*128,0);
    nread=fread(EPP->access_table,1,(EPP->lr-EPP->fr+1)*2,EPP->fil);
    if (nread!=(EPP->lr-EPP->fr+1)*2)
	  /*if not right amount,assume bad*/
	    EPP->access_ptr=0;
    if (REVERSE) swap2((short*)EPP->access_table,EPP->lr-EPP->fr+1);
    fseek(EPP->fil,now,0);
  }
  if (EPP->access_ptr==0) {  /*some kind of bad table,just count*/
    if (row<EPP->row_cnt+EPP->fr) {
      if (!epprewind(EPP)) return 0;
    }
    for (r=EPP->fr+EPP->row_cnt; r<row; r++) {
      if (!get_row(EPP)) return 0;
    }
    return 1;
  }
  z=128;   /*header*/
  for (r=0; r<row-EPP->fr; r++)
    z+=EPP->access_table[r];
  if (feof(EPP->fil) && EPP->lastbyte!=4096)
    now=ftell(EPP->fil)-EPP->lastbyte+256;
  else
    now=ftell(EPP->fil)-4096;
  if ((unsigned long)(z-now)<3840) {
    EPP->fp=EPP->fptr+z-now+256;
    return 1;
  }
  EPP->fp=EPP->fptr+(z & 127)+256;
  fseek(EPP->fil,z & 0xffffff80,0);
  nread=fread(&EPP->fptr[256],1,4096,EPP->fil);
  if (nread==4096)
    EPP->lastbyte=4096;
  else
    EPP->lastbyte=nread+256;
  return 1;
}


char eppclose(eppfile *EPP)
{
  char badfile,eightbit;

  badfile=0;
  eightbit=0;
  if (EPP->access_table!=NULL) free(EPP->saved);
  if (EPP->rptr!=NULL) free(EPP->rptr);
  free(EPP->fptr);
  fclose(EPP->fil);
  return 1;
}

#ifdef LEGENDS
char legfile::reset(void)
{
  char j,k;
  long v;
  char s[100],*s2;
  legend *p,*p1,*p2;

  root=NULL;
  defaultleg.level=1;
  defaultleg.npass=1;
  memcpy(defaultleg.printchar,"    ",4);
  defaultleg.next=NULL;
  if (*filname=='\0') return 0;
#ifdef __BORLANDC__ 
  _fullpath(s,filname,80);
  strcpy(filname,s);
#else
  if (*filname!='/') {
    getcwd(s,80);
    strcat(s,"/");
    strcat(s,filname);
    strcpy(filname,s);
  }
#endif
  addExt(filname,".leg",0);
  tfil=fopen(filname,"r");
  if (tfil==NULL) return 0;
  while (!feof(tfil)) {
    fgets(s,100,tfil);
    if (strchr(s,'\n')!=NULL) *strchr(s,'\n')=0;
    j=sscanf(s,"%ld",&v);
    if (v<0 || j==0) continue;
    s2=s;
    while (*s2==' ') s2++;
    while (*s2 && *s2!=' ') s2++;
    while (*s2==' ') s2++;
    p=(legend *)malloc(sizeof(legend)-255+strlen(s2));
    strcpy(p->legtext,s2);
    p->level=v;
    p->npass=0;
    memcpy(p->printchar,"    ",4);
    if (root==NULL) {
      root=p;
      root->next=NULL;
      continue;
    }
    p1=root;
    while (p1!=NULL && p1->level<v) {
      p2=p1;
      p1=p1->next;
    }
    if (p1==NULL) {   /*already there,keep first line*/
      p2->next=p;
      p->next=NULL;
      continue;
    }
    if (p1->level==v) {
      free(p);
      continue;
    }
    if (p1==root) {
      p->next=root;
      root=p;
    } else {
      p->next=p2->next;
      p2->next=p;
    }
  }
  fclose(tfil);
  return (root!=NULL);

}


legend *legfile::getptr(epplev v)
{
  legend *p;

  p=root;
  while (p!=NULL && p->level<v)
    p=p->next;
  if (p==NULL || p->level!=v) {
    defaultleg.printchar[0]=v % 78+'0';
    *defaultleg.legtext='\0';
    return (&defaultleg);
  } else
    return p;
}


char legfile::close(void)
{
  char n;
  legend *p1,*p2;

  for (n=0; n<=14; n++) if (this==open_fibs[n]) open_fibs[n]=NULL;
  p1=root;
  while (p1!=NULL) {
    p2=p1->next;
    free(p1);
    p1=p2;
  }
  return 1;
}
#endif

char clrreset(clrfile *CLR)
{
  int valu;
  int n,r,g,b,sz;
  clrTag tmp[300];
  char STR1[80];

  CLR->lastColor=0;
  strcpy(strrchr(CLR->filname,'.'),".clr");
  CLR->fil=fopen(CLR->filname,"r");
  if (CLR->fil==NULL) return 0;
  memset(tmp,0,sizeof(tmp));
  sz=sizeof(tmp)/300;

  while (!feof(CLR->fil)) {
    fgets(STR1,80,CLR->fil);
    sscanf(STR1,"%d%d%d%d",&valu,&r,&g,&b);
    n=CLR->lastColor-1;
    while (n>=0 && tmp[n].eppval>valu) {
      tmp[n+1]=tmp[n];
      n--;
    }
    if (r>=1000) r=999;
    if (g>=1000) g=999;
    if (b>=1000) b=999;
    tmp[n+1].color.red=r*32/125;
    tmp[n+1].color.green=g*32/125;
    tmp[n+1].color.blue=b*32/125;
    tmp[n+1].eppval=valu;
    CLR->lastColor++;   /*?? check overflow*/
  }
  fclose(CLR->fil);
  CLR->clrTbl=(clrTag *)malloc(CLR->lastColor*sz);
  memmove(CLR->clrTbl,tmp,CLR->lastColor*sz);
  return (CLR->lastColor!=0);
}


void clrget(clrfile *CLR,epplev n,TRGB *color)
{
  int ndx;

  ndx=0;   /**** use binary search for efficiency,later*/
  while (ndx<CLR->lastColor && CLR->clrTbl[ndx].eppval<n)
    ndx++;
  if (ndx!=CLR->lastColor && CLR->clrTbl[ndx].eppval==n)
  {   /*lousy choice of names*/
    *color=CLR->clrTbl[ndx].color;
    return;
  }
  color->red=0;   /*paint it black...*/
  color->green=0;
  color->blue=0;
}

char clrclose(clrfile *CLR)
{
  free(CLR->clrTbl);
  return 1;
}

